pkg/session/state.go

```diff
@@ -0,0 +1,293 @@
+				if c, has := spt.spm[id]; has && c.concurrency > 0 {
+					c.concurrency--
+					spt.spm[id] = c
+				} else {
+					return fmt.Errorf("internal error, peer.ID not properly double-recorded %s", retrievalId)
+				}
+				break
+			}
+		}
+		if !foundSp {
+			return fmt.Errorf("internal error, no such storage provider %s for retrieval %s", storageProviderId, retrievalId)
+		}
+		spt.arm[retrievalId] = ar
+	} else {
+		return fmt.Errorf("internal error, no such active retrieval %s", retrievalId)
+	}
+
+	return nil
+}
+
+func (spt *SessionState) RecordFailure(storageProviderId peer.ID, retrievalId types.RetrievalID) error {
+	// remove from this retrieval to free up the SP to be tried again for a future retrieval
+	if err := spt.RemoveFromRetrieval(storageProviderId, retrievalId); err != nil {
+		return err
+	}
+
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
+
+	status := spt.spm[storageProviderId]
+
+	// Filter out expired history
+	n := 0
+	for _, failure := range status.failures {
+		if time.Since(failure) <= spt.config.FailureHistoryDuration {
+			status.failures[n] = failure
+			n++
+		}
+	}
+	status.failures = status.failures[:n]
+
+	// Add new failure to history
+	status.failures = append(status.failures, time.Now())
+
+	// Decide whether to suspend miner
+	if len(status.failures) > int(spt.config.MaxFailuresBeforeSuspend) {
+		status.suspensionStart = time.Now()
+		status.suspensionDuration = spt.config.SuspensionDuration
+	}
+
+	if status.isSuspended() {
+		logger.Warnf(
+			"Suspending storage provider for %s after %d failures within %s: %s",
+			status.suspensionDuration,
+			len(status.failures),
+			time.Since(status.failures[0]),
+			storageProviderId,
+		)
+	}
+
+	// Write updated status back to map
+	spt.spm[storageProviderId] = status
+	return nil
+}
+
+func (spt *SessionState) RegisterConnectTime(storageProviderId peer.ID, current time.Duration) {
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
+
+	status := spt.spm[storageProviderId]
+	// EMA of connect time
+	if !status.connectTimeInitialized {
+		status.connectTimeInitialized = true
+		status.connectTime = current
+	} else {
+		status.connectTime = time.Duration((1-spt.config.ConnectTimeAlpha)*float64(current) + spt.config.ConnectTimeAlpha*float64(status.connectTime))
+	}
+	spt.spm[storageProviderId] = status
+}
+
+func (spt *SessionState) CompareStorageProviders(protocol multicodec.Code, a, b peer.ID, mda, mdb metadata.Protocol) bool {
+	if protocol == multicodec.TransportGraphsyncFilecoinv1 {
+		gsmda := mda.(*metadata.GraphsyncFilecoinV1)
+		gsmdb := mdb.(*metadata.GraphsyncFilecoinV1)
+		// prioritize verified deals over not verified deals
+		if gsmda.VerifiedDeal != gsmdb.VerifiedDeal {
+			return gsmda.VerifiedDeal
+		}
+
+		// prioritize fast retrievel over not fast retrieval
+		if gsmda.FastRetrieval != gsmdb.FastRetrieval {
+			return gsmda.FastRetrieval
+		}
+	}
+
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
+	sa := spt.spm[a]
+	sb := spt.spm[b]
+
```

Comment:
seems like we want to bring failures into this equation pretty soon in a subsequent PR